framework/src/constraints/ADPeriodicSegmentalConstraint.C

```diff
@@ -0,0 +1,89 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ADPeriodicSegmentalConstraint.h"
+
+registerMooseObject("MooseApp", ADPeriodicSegmentalConstraint);
+
+InputParameters
+ADPeriodicSegmentalConstraint::validParams()
```

Comment:
I think this would mean making Generic versions of the Scalar Base classes, right? When I look at 
https://mooseframework.inl.gov/docs/doxygen/moose/GenericKernel_8h_source.html I see a bunch of "using" statements. Is there some rule to figure out what lines I need, or just keep adding variables until I don't get error messages? I.e. is there a forum post or someone to ask about this?

Reply:
Okay, so I decided to just take a stab at it by looking at the analogy in other files. I added the stub Qp methods for the AD base class so that they override, and made one test case for a Kernel with scalars. It compiles and it seems to run the same way as the current classes that I copied from.
Take a look and see if there are any extraneous declarations I made in the using header, and any other feedback.

Reply:
Unfortunately, the generic version of the mortar objects will be more difficult because of the special effects in the AD version of these source files. Namely, the code actually isn't duplicated but is distinct. Due to the adding of AD derivatives to vectors and with the pointers to _phi from different types. So, I'll either need help with this, or we just stick with the Kernel example to illustrate how this would be done.

Reply:
I'm impressed you took a stab at any of them. Let's not worry about the mortar objects.